AGENT-1193: Add --mirror-path flag to support pre-mirrored images

rwsu opened the pull request:
**Adds support for using pre-mirrored OCP release images instead of running oc-mirror during the build process.** This is useful when images have already been mirrored in a separate step, avoiding redundant mirroring operations.

## Changes

- **Add \`--mirror-path\` flag to the build command** — Specifies the location of pre-mirrored registry data. When provided, the appliance builder skips running \`oc-mirror\` and uses the pre-mirrored images directly.

- **Fix image reference parsing for registries with ports** — When a custom release URL contains a tag (e.g. \`registry:5000/img:tag\`), appending a digest directly would produce a \`tag@digest\` reference rejected by image validation. Uses \`LastIndex\` to correctly strip the tag while preserving the registry port.

- **Refactor mirror-path handling** — Extract registry data copy logic into \`copyMirrorRegistryData()\` and reorder \`mirrorImages()\` condition so the normal flow comes first.

---

**Assisted-by:** Claude Sonnet 4.6 <noreply@anthropic.com>

rwsu committed:
Add mirror-path support and fix custom registry handling

Add support for using pre-mirrored images from oc-mirror workspace by
specifying the mirror path in appliance-config.yaml. When mirrorPath is
configured, the appliance skips running oc-mirror and uses the pre-mirrored
registry data directly.

Also fixes issues when using custom registries (non-quay.io) where oc adm
release info may return incomplete image references missing port and
repository path. Adds IDMS entries to ensure the cluster redirects pulls
from custom registries to the appliance's internal registry.

Changes:
- Add MirrorPath field to top level of ApplianceConfig in types
- Update appliance-config.yaml template to document the new field
- Update code to read mirrorPath from ApplianceConfig.Config.MirrorPath
- Add validateMirrorPath() to ApplianceConfig with comprehensive validation
- Use pre-mirrored images when mirrorPath is provided
- Add fixImageReference() to repair incomplete image refs from oc commands
  Example: registry.example.com@sha256:... becomes
           registry.example.com:5000/repo/image@sha256:...
- Add addLocalRegistryIDMS() to generate IDMS for custom registry mirrors
- Fix release image tag parsing to preserve port in registry URLs

The mirror-path can be specified in appliance-config.yaml as:
  mirrorPath: /path/to/oc-mirror/workspace

Assisted-by: Claude Sonnet 4.5 <noreply@anthropic.com>

Add debug logging for IsLiveISO tracking

rwsu committed:
Add --registry-config to oc commands for custom registry authentication

Pass --registry-config ~/.docker/config.json to all oc commands
(release info, release extract, oc mirror) and --authfile to skopeo.
This allows the appliance builder to authenticate against private
registries such as the one used in dev-scripts
(virthost.ostest.test.metalkube.org:5000).

The pull secret from appliance-config is written to ~/.docker/config.json
(existing behavior) so no separate credential management is needed.

Assisted-by: Claude Sonnet 4.6 <noreply@anthropic.com>

danielerez commented on pkg/asset/config/appliance_config.go:
```diff
@@ -553,23 +566,87 @@ func (a *ApplianceConfig) validatePinnedImageSet() error {
-	configPath := filepath.Join(homeDir, ".docker", "config.json")
-	if err = os.MkdirAll(filepath.Dir(configPath), os.ModePerm); err != nil {
-		return err
+	// Write pull secret to temp file
```
why is it needed? we already store it in .docker/config.json, where oc can use it.
I mean, why not to keep lines 564-565?

rwsu commented on cmd/build.go:
```diff
@@ -28,6 +28,12 @@ var (
+	// is reused throughout command execution. This prevents EnvConfig from
+	// being regenerated with default values when setupApplianceConfig() fetches
+	// ApplianceConfig, which would overwrite runtime flags like IsLiveISO.
+	assetStoreInstance asset.Store
```
Yes, I retested and it is not needed. I will remove.

rwsu commented on pkg/asset/config/appliance_config.go:
```diff
@@ -153,6 +154,12 @@ pullSecret: pull-secret
+# When provided, skips image mirroring and uses the pre-mirrored registry data.
+# The path should point to an oc-mirror workspace directory containing a 'data' subdirectory.
+# [Optional]
+# mirrorPath: /path/to/mirror/workspace
```
Done.

danielerez commented on pkg/asset/data/data_iso.go:
```diff
@@ -122,7 +123,39 @@ func (a *DataISO) Generate(dependencies asset.Parents) error {
 	imageGen := genisoimage.NewGenIsoImage(nil)
-	if err = imageGen.GenerateImage(envConfig.CacheDir, dataIsoName, filepath.Join(envConfig.TempDir, dataDir), dataVolumeName); err != nil {
+
+	// When mirror-path is provided, copy the Docker registry data from mirror-path/data
```
this Generate func is getting a bit too long:)
can you extract the new logic to another func?

rwsu replied:
Yes, the new logic is now in the copyMirrorRegistryData func.

rwsu committed:
Fix tag stripping for release image URLs with registry port

strings.Split(image, ":")[0] incorrectly strips the registry port when
the release image URL contains one. For example, with the dev-scripts
local registry virthost.ostest.test.metalkube.org:5000/openshift/release-images:tag,
Split on ":" takes only "virthost.ostest.test.metalkube.org", losing the
port and path entirely.

Use LastIndex to locate the tag colon after the last "/" so the port is
preserved, producing the correct digest reference:
  virthost.ostest.test.metalkube.org:5000/openshift/release-images@sha256:...

Extract the logic into appendDigest() with a unit test.

Assisted-by: Claude Sonnet 4.6 <noreply@anthropic.com>

rwsu committed:
Refactor mirror-path handling in mirrorImages and DataISO

- Swap mirrorPath condition in mirrorImages() so normal flow comes first
- Extract mirror-path registry data copy logic into copyMirrorRegistryData()

Assisted-by: Claude Sonnet 4.6 <noreply@anthropic.com>

rwsu commented on pkg/asset/config/appliance_config.go:
```diff
@@ -361,24 +368,25 @@ func (a *ApplianceConfig) GetRelease() (string, string, error) {
 		}
 		releaseVersion = strings.Trim(releaseVersion, "'")
-		logrus.Debugf("Release version: %s", releaseVersion)
+		logrus.Debugf("Got release version: '%s'", releaseVersion)
```
I've reverted it back to the original message.

rwsu commented on pkg/release/release.go:
```diff
@@ -138,35 +138,49 @@ func (r *release) execute(command string) (string, error) {
+		logrus.Infof("Using pre-mirrored images from: %s", mirrorPath)
+		tempDir = mirrorPath
+	} else {
+		// Normal mirroring flow - run oc-mirror
```
Yes, done.

rwsu commented on pkg/asset/config/appliance_config.go:
```diff
@@ -361,24 +368,25 @@ func (a *ApplianceConfig) GetRelease() (string, string, error) {
 			}
 			releaseDigest = strings.Trim(releaseDigest, "'")
-			releaseImage = fmt.Sprintf("%s@%s", strings.Split(releaseImage, ":")[0], releaseDigest)
+			releaseImage = fmt.Sprintf("%s@%s", releaseImage, releaseDigest)
```
I've create a unit test to verify it doesn't break images without port.

danielerez commented on pkg/asset/config/appliance_config.go:
```diff
@@ -47,8 +47,8 @@ const (
 	// Release
-	templateGetVersion = "oc adm release info %s -o template --template '{{.metadata.version}}'"
-	templateGetDigest  = "oc adm release info %s -o template --template '{{.digest}}'"
+	templateGetVersion = "oc adm release info --registry-config %s %s -o template --template '{{.metadata.version}}'"
```
Why is it mandatory to explicitly pass the pull secret?
I.e. we already store it in the default location: https://github.com/openshift/appliance/blob/75d95da1beae653e1ab865d5e88cc355f3af65f4/pkg/asset/config/appliance_config.go#L556

rwsu replied:
It is not needed. See [comment](https://github.com/openshift/appliance/pull/634#discussion_r3115345808).

danielerez commented on pkg/release/release.go:
```diff
@@ -32,10 +32,10 @@ const (
-	templateExtractCmd   = "oc adm release extract --command=%s --to=%s %s"
-	templateImageExtract = "oc image extract --path %s:%s --confirm %s"
-	ocMirror             = "oc mirror --v2 --config=%s docker://127.0.0.1:%d --workspace=file://%s --src-tls-verify=false --dest-tls-verify=false --parallel-images=4 --parallel-layers=4 --retry-times=5 --ignore-release-signature"
+	templateGetImage     = "oc adm release info --registry-config %s --image-for=%s %s"
```
Same here about the pull secret. Which scenario requires passing the pull secret explicitly now?

rwsu replied:
Good catch. You're correct — oc and skopeo already pick up credentials from ~/.docker/config.json automatically, which is where storePullSecret() already writes them. We've removed --registry-config from all oc commands and --authfile from skopeo — they  were redundant. Verified the ISO build and cluster install succeed without them.

bfournie commented:
With this PR I've been able to use a local image (machine-config-operator in this case) and successfully install a cluster via the dev-scripts UI-driven installation, confirming that my local changes were picked up. 
Changes look good to me but will wait for all of Daniel's questions to be answered.

rwsu committed:
Remove explicit --registry-config and --authfile flags from oc and skopeo

oc and skopeo automatically pick up credentials from ~/.docker/config.json,
which is already written by storePullSecret(). Passing the flags explicitly
is redundant and was only introduced when the pull secret was temporarily
stored in a separate temp file.

Assisted-by: Claude Sonnet 4.6 <noreply@anthropic.com>

bfournie commented:
Testing the PR, one thing we've noticed that we're missing in the mirrored data is` openshift/release-bundles`, and the operator data. We can probably work around the missing operator data by adding these in dev-scripts but it seems that the `openshift/release-bundles` should be added here if possible.

This is what the local catalog looks like on an installed cluster (using a local mco image):
```
​$ curl -sk https://192.168.111.20:22625/v2/_catalog | jq -r
{
  "repositories": [
    "localimages/agent-installer-utils",
    "localimages/machine-config-operator",
    "openshift/release",
    "openshift/release-images",
    "rhel9/support-tools",
    "ubi8/ubi"
  ]
}
```
Compared to a system installed without the local mirror:
```
{
  "repositories": [
   <operator specific repos>
    "openshift/release",
    "openshift/release-bundles",
    "openshift/release-images",
    "rhel9/support-tools",
  ]
}
```

So we're missing `openshift/release-bundles` and the operator info.

rwsu committed:
Fix mirror-path registry data path and missing release-bundles

Two related bugs in the --mirror-path flow:

1. copyMirrorRegistryData was called after bundle.Push(), causing the
   mirror-path docker data to overwrite the release-bundles image that
   was just pushed to the registry. Move the copy before starting the
   registry so bundle.Push() adds release-bundles on top of the
   pre-populated mirror data.

2. Moving the copy earlier exposed a path mismatch: registryDir from
   GetRegistryDataPath() was mangled (e.g. /assetstemp/data instead of
   /assets/temp/data). Switch to dataDirPath (filepath.Join(envConfig.TempDir,
   dataDir)) throughout, consistent with how CopyRegistryImageIfNeeded
   computes paths.

Assisted-by: Claude Sonnet 4.6 <noreply@anthropic.com>

rwsu commented:
> So we're missing `openshift/release-bundles` and the operator info.

@bfournie The missing release-bundles should be fixed by [this commit](https://github.com/openshift/appliance/pull/634/commits/a16c7319c82f2a655a8f6a941a0bfd869a60224e) the missing operators can be added through [this dev-scripts PR](https://github.com/openshift-metal3/dev-scripts/pull/1883).

danielerez commented:
/retest

rwsu commented:
/retest-required

rwsu commented:
/test e2e-agent-compact-ipv4-iso-no-registry

bfournie commented:
With the latest change I confirmed that the release-bundles are now available
```
curl -sk https://192.168.111.20:22625/v2/_catalog?n=1000 | jq -r
{
  "repositories": [
    "localimages/agent-installer-utils",
    "localimages/machine-config-operator",
    "openshift/release",
    "openshift/release-bundles",
    "openshift/release-images",
    "rhel9/support-tools",
    "ubi8/ubi"
  ]
}
$ curl -sk https://localhost:22625/v2/openshift/release-bundles/tags/list
{"name":"openshift/release-bundles","tags":["ocp-release-bundle-4.23.0-0.ci-2026-04-20-103815"]}
```

bfournie commented:
/lgtm
